sdk/nodejs/serverless/function.ts

```diff
@@ -0,0 +1,184 @@
+
+/**
+ * Handler is the signature for a serverless function.
+ */
+export type Handler = (request: any, response: any) => any;
+
+/**
+ * FunctionOptions provides configuration options for the serverless Function.
+ */
+export interface FunctionOptions {
+    /**
+     * Memory (in MB), available to the function. Default value is 256MB. Allowed values are: 128MB, 256MB, 512MB, 1024MB, and 2048MB.
+     */
+    readonly availableMemoryMb?: pulumi.Input<number>;
+    /**
+     * Description of the function.
+     */
+    readonly description?: pulumi.Input<string>;
+    /**
+     * A set of key/value label pairs to assign to the function.
+     */
+    readonly labels?: pulumi.Input<{
+        [key: string]: any;
+    }>;
+    /**
+     * Project of the function. If it is not provided, the provider project is used.
+     */
+    readonly project?: pulumi.Input<string>;
+    /**
+     * Region of function. Currently can be only "us-central1". If it is not provided, the provider region is used.
+     */
+    readonly region?: pulumi.Input<string>;
+    /**
+     * Timeout (in seconds) for the function. Default value is 60 seconds. Cannot be more than 540 seconds.
+     */
+    readonly timeout?: pulumi.Input<number>;        
+   /**
+    * The packages relative to the program folder to not include the Function upload. This can be
+    * used to override the default serialization logic that includes all packages referenced by
+    * project.json (except @pulumi packages).  Default is `[]`.
+    */
+   excludePackages?: string[];
+}
+
+/**
+ * Function is a higher-level API for creating and managing GCP Cloud Function resources implemented
+ * by a given handler.
+ */
+export class Function extends pulumi.ComponentResource {
+    public readonly options: FunctionOptions;
+    public readonly function: cloudfunctions.Function;
+    public readonly bucket: storage.Bucket;
+    public readonly bucketObject: storage.BucketObject;
+
+    constructor(name: string,
+        options: FunctionOptions,
+        func: Handler,
+        opts?: pulumi.ResourceOptions,
+        serialize?: (obj: any) => boolean) {
+        if (!name) {
+            throw new Error("Missing required resource name");
+        }
+        if (!func) {
+            throw new Error("Missing required function callback");
+        }
+
+        super("gcp:serverless:Function", name, { options: options }, opts);
+
+        // Now compile the function text into an asset we can use to create the function. Note: to
+        // prevent a circularity/deadlock, we list this Function object as something that the
+        // serialized closure cannot reference.
+        serialize = serialize || (_ => true);
+        const finalSerialize = (o: any) => {
+            return serialize(o) && o !== this;
+        }
+
+        const handlerName = "handler";
+        const serializedFileNameNoExtension = "index";
+
+        let closure = pulumi.runtime.serializeFunction(func, {
```

Comment:
If we were correctly linting this, we'd enforce that `const` be used when possible.  I think we don't have linting set up correctly here yet - but we should still use `const` consistently.